src/server.c

```diff
@@ -1736,27 +1736,61 @@ int clientsCronTrackExpansiveClients(client *c, int time_idx) {
     return 0; /* This function never terminates the client. */
 }
 
-/* Iterating all the clients in getMemoryOverheadData() is too slow and
- * in turn would make the INFO command too slow. So we perform this
- * computation incrementally and track the (not instantaneous but updated
- * to the second) total memory used by clients using clinetsCron() in
- * a more incremental way (depending on server.hz). */
-int clientsCronTrackClientsMemUsage(client *c) {
-    size_t mem = 0;
+clientMemUsageBucket *getMemUsageBucket(size_t mem) {
+    int size_in_bits = 8*(int)sizeof(mem);
+    int clz = mem > 0 ? __builtin_clzl(mem) : size_in_bits;
+    int bucket_idx = size_in_bits - clz;
+    if (bucket_idx >= CLIENT_MEM_USAGE_BUCKETS)
+        bucket_idx = CLIENT_MEM_USAGE_BUCKETS-1;
+    return &server.client_mem_usage_buckets[bucket_idx];
+}
+
+/* This is called both on explicit clients when something changed their buffers,
+ * so we can track clients' memory and enforce clients maxmemory in real time
+ * and also from the clietsCron so we have updated stats for non
+ * CLIENT_TYPE_NORMAL/PUBSUB clients.
```

Comment:
are we sure we really need this in cron too?

Reply:
We need it in the cron in order to make sure the other (non client eviction) stats are also updated.
There are also edge cases where a normal client has large buffers but is marked as `CLIENT_NO_EVICT` and has no activity on it anymore. Then the user turns off this flag but no new commands are executed and the write event never happens. In this case we need the cron in order to put the client into some bucket.

Reply:
maybe drop a hint in the comment about this for future reference..
i.e. you already did, but i think it may be better to improve it.

Reply:
Done.